### ISSUE TITLE
Switch Roc to passed in allocators

### ISSUE THREAD

**Issue by bhansconnect** [roc-lang/roc#6382]
Context: https://roc.zulipchat.com/#narrow/stream/304641-ideas/topic/add.20an.20.60init.60.20function.20to.20.60roc.60.20code

When calling Roc, a language should pass in a pointer to a struct that is essentially defined as:
```rust
// Fields would be sorted alphabetically
struct Allocator {
    context: *mut c_void,
    alloc:
        *const extern "C" fn(context: *mut c_void, size: usize, alignment: u32) -> *mut c_void,
    realloc: *const extern "C" fn(
        context: *mut c_void,
        c_ptr: *mut c_void,
        new_size: usize,
        _old_size: usize,
        _alignment: u32,
    ) -> *mut c_void,
    // ... other allocator functions

    // Until effect interpreters are a thing, include all `roc_fx_*` functions
}
```

Roc will implicitly pass the pointer to this struct into all functions (future optimization, just those that can allocate). Whenever any of these functions are needed, it will load the pointer from the struct and call them.

For the most part, the api of each of these calls is the same as before, but with an extra `*mut c_void` context parameter. The context is added by the host to the struct and gets passed into each call. This enables the host to easily store context local values. For example. this makes it trivial for each async thread to have their own memory arena. Just set the context to a pointer to the local arena.

-----

This has two major gains from current roc:
1. It fixes some major linking pain points caused by roc calling functions in the host (especially as a shared library plugin).
2. It gives the host much cleaner and nicer control over allocations.

As a third minor short term bonus, the host could include extra data in the context that would be used by effects. Once we have effect interpreters, that won't matter anymore though.

**Comment by ostcar** [comment]
This sound wonderful.

When you are implementing this, could you also consider changing the signature of `roc_free` to have an `size` argument?

**Comment by bhansconnect** [comment]
The `size` argument would be more complicated. For example, a seamless slice does not know the capacity of the underlying allocation. (not sure if there is anything else, may just be seamless slices that are dynamic in a complex way like this)

My gut feeling is that we should leave the size as a problem for the allocator. Most allocators already track the size, so us storing it for dynamic allocation is just a waste of memory. For any allocator that doesn't already track the size, it is trivial to make every allocation slightly larger such that the size can be stored before the allocation.

**Comment by ostcar** [comment]
In this case, you could remove the `_old_size` argument from `realloc`.

**Comment by bhansconnect** [comment]
Yeah, probably should remove it. Though we are technically guaranteed to know it in that case.
